### ISSUE TITLE
Finalize migration from Firebase 2.4.0 to 3.x

### ISSUE THREAD

**Issue by Cyclodex** [Cyclodex/cyclone#5]
Currently Cyclone uses a mix of Firebase, for the new authentication.
The base of the software is however using the old 2.4.0 version still, as well as some dependencies as angularfire etc.
Update everything to have only 1 version of firebase

**Comment by Cyclodex** [comment]
@pfiaux I am working on this, but still having troubles to get the authentication state into the connection Controller. Will continue and create a branch tomorrow/later.

**Comment by pfiaux** [comment]
You can probably create a service, say `cycloneBackendService` or `cycloneAuthService` there you can
- Have a get authentication state method
- Store the state when connected so you can reuse it
- You can then inject the `cycloneBackendService` to any controller and access the data anywhere.
  - The goal would be that controllers don't use firebase at all, they only know about the services.

This is a good example I think: https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md#separate-data-calls

So something like:

```
angular.module('cycloneApp').factory('dataService', ['$firebaseAuth', function($firebaseAuth) {
    function DataService() {
        this._state = false;
    }

    DataService.prototype.getAuthenticationState = function(asdf) {
        // If we already have it
        if (this._state) {
            return this._state;
        }
        else {
            // TODO i'm just guessing I dont know fire base:
            this._state = $firebaseAuth.getState();
            return this._state;
        }
    }

    return new DataService();
}]);
```

This is just an example you probably need to decide if it's a promise or not, if you get it once on initialization or only once you need it and if you need to handle timeout, reconnection and more. Then in your controller you can do something like:

```
dataService.getAuthenticationState().then(...);
```

**Comment by Cyclodex** [comment]
Thanks!
I will commit this soon. Its great to improve the code now step by step, thanks for your help and hints and reviews! 👍 🎉 

# Auth service

I am looking at angularfire examples directly to see how this should be done.
- https://github.com/firebase/angularfire/blob/master/docs/guide/user-auth.md#retrieving-authentication-state

# How to load all the libraries

I am thinking of the libraries as well, should we try to load them all with npm or so ?
- https://github.com/firebase/angularfire#downloading-angularfire

Would probably also have an impact on the webpack build...

**Comment by Cyclodex** [comment]
@pfiaux I looked into it and see an example which uses then the $scope to use the authentication, but I think we don't need this, at least not yet. It should be fine to use directly the "Auth" variable right?

The example showed:

```
$scope.auth = Auth;
$scope.auth.$onAuthStateChanged(function(user) {
```

I think this should be enough, if we are not using it multiple times within a controller.

```
Auth.$onAuthStateChanged(function(user) {
```

PS: I know we have to change a lot more here... Lets go step by step :neckbeard:

**Comment by pfiaux** [comment]
I quickly looked at where you use Auth in the code, note that `$onAuthStateChanged` can be called multiple times, so the function you put inside it will be called everytime the AuthState changes.

But yes it's better to use Auth, you want to avoid using $scope as much as possible.

If you only need it once you have to store the unsubscribe function and call it: https://github.com/firebase/angularfire/blob/master/docs/reference.md#onauthstatechangedcallback-context or you can keep it so if the user changes it will update the controller but then you probably want to unregister the call back when we leave the controller like:

```
$scope.$on("$destroy", function() {
    offAuth();
});
```

Or like we said if you only want this once and you wrap `$firebaseAuth()` inside of your service to avoid exposing it: 

```
// Inside of your Auth service:
function getUser() {
    return $q(function(resolve, reject) {
        var unsubscribe = auth.$onAuthStateChanged(function(user) {
            if (user) {
                resolve(user);
            }
            // We're not logged in reject:
            else {
                reject(new Error('No user signed in'));
            }
        });
    });
}
```

But it might not be worth refactoring if it already works. I think it would actually be better with listerners and call backs so you can move most of the firebase code out of the controller. Maybe an event cleaner way to handle it would be some thing like:
https://www.npmjs.com/package/rx-firebase but you need to get into reactive programming first ;). So like you said one step at a time maybe later :)

**Comment by Cyclodex** [comment]
Okey, lets go further in a separate step. I also think this probably has to deal more with route etc. which we also can improve at some point.
I am closing the migration to Firebase 3.x as I tested today the whole day and it seems to work.

**Comment by Cyclodex** [comment]
This is fixed and merged into master in #19
